Fix iOS 13+ multi-scene window handling and thread safety

AdrianCurtin opened the pull request:
## Summary

This PR consolidates multiple bug fixes to improve iOS 13+ compatibility, thread safety, and defensive nil handling.

### iOS 13+ Multi-Scene Support
- Rewrote `+mainWindow` to properly enumerate connected scenes and find the key window
- Added CarPlay filtering via `UIWindowSceneSessionRoleApplication` check
- Updated `-frontWindow` and `-visibleKeyboardHeight` for scene-based window discovery
- Added proper fallbacks for pre-iOS 13 and edge cases where no active scene exists

### Thread Safety
- `+popActivity` now checks `[NSThread isMainThread]` and dispatches to main queue when needed
- `-showProgress:status:` optimized to execute synchronously when already on main thread

### Memory Management
- Fixed weak/strong dance in `-showProgress:status:`, `-showImage:status:duration:`, and `-dismissWithDelay:completion:` to consistently use `strongSelf`

### Nil Safety
- Added nil check in `+imageBundle` before calling `bundleWithURL:`
- Added `containerView.window` checks in `-updateViewHierarchy` and `-moveToPoint:rotateAngle:`
- Added `window && window.rootViewController` check before adding HUD to window
- Added nil check for `rootController` before calling `setNeedsStatusBarAppearanceUpdate`

### Other
- Fixed umbrella header warning by removing internal header symlinks
- Replaced `performSelector:` with proper protocol conformance for `UIWindowSceneDelegate`

## Closes

- Closes #858 - Error in popActivity implementation
- Closes #1077 - Umbrella header found (SPM)
- Closes #1084 - SVProgressHUD displays in top left corner
- Closes #1113 - Main Window acquisition is inaccurate
- Closes #1121 - There are still cases where the HUD is not displayed
- Closes #1144 - Symlink missing for SVIndefiniteAnimatedView
- Closes #1146 - SVProgressHUD Crash on [NSBundle bundleWithURL:]
- Closes #1150 - Crash happened when CarPlay Scene connected
- Closes #1153 - [SVProgressHUD showWithStatus:] doesn't work on iOS 26 most of the time
- Closes #1154 - SVProgressHUD crashes in applications with CarPlay support
- Closes #1155 - Crash in dismiss animation completion when rootViewController is nil
- Closes #1157 - Crash in updateViewHierarchy when window has no rootViewController

## Merges

- #927 - popActivity added to operationQueue
- #1087 - fix async issues (iOS 15)
- #1114 - fix mainWindow acquisition is inaccurate
- #1118 - Add extra check for CarPlay window scene
- #1131 - Fix warning Umbrella header for module
- #1152 - Adapt to UIKit Scene-Based Life Cycle (iOS 26 Migration)

ibarisic05 committed:
popActivity added to operationQueue

jls21stuzo committed:
fix asyc issues (ios 15)

CloudlessMoon committed:
fix mainWindow acquisition is inaccurate https://github.com/SVProgressHUD/SVProgressHUD/issues/1113

datwelk committed:
add check for carplay window

datwelk committed:
cleanup

Commit:
Fix warning Umbrella header for module

Update CHANGELOG.md

Fix warning Umbrella header for module

Commit:
Fix: Adapt window retrieval for multi-scene life cycle (iOS26: https://developer.apple.com/documentation/technotes/tn3187-migrating-to-the-uikit-scene-based-life-cycle)

AdrianCurtin committed:
Fix crash when rootViewController is nil

Added a nil check before calling setNeedsStatusBarAppearanceUpdate on rootViewController to prevent potential crashes when rootViewController is not set.

AdrianCurtin committed:
Add nil check in updateViewHierarchy before adding to window

AdrianCurtin committed:
Merge PR #1118: Add extra check for CarPlay window scene

AdrianCurtin committed:
Merge PR #1158: Add nil check in updateViewHierarchy before adding to window

AdrianCurtin committed:
Merge PR #1152: Adapt to UIKit Scene-Based Life Cycle (iOS 26 Migration)

AdrianCurtin committed:
Merge PR #1087: fix async issues (iOS 15)

AdrianCurtin committed:
Merge PR #927: popActivity added to operationQueue

AdrianCurtin committed:
Merge PR #1131: Fix warning Umbrella header for module

AdrianCurtin committed:
Merge PR #1114: fix mainWindow acquisition is inaccurate

AdrianCurtin committed:
Fix crash in imageBundle when bundle URL is nil

Adds nil check before calling bundleWithURL: to prevent crash
when SVProgressHUD.bundle cannot be found.

Fixes #1146

AdrianCurtin committed:
Fix HUD not displaying when containerView is no longer in hierarchy

Check if containerView.window exists before using containerView.
Falls back to frontWindow if containerView was deallocated.

Fixes #1121

AdrianCurtin committed:
Fix window access and thread handling in SVProgressHUD

Refactored window retrieval to use protocol conformance and property access for UIWindowSceneDelegate and UIApplicationDelegate. Improved thread safety in popActivity by checking for main thread before executing. Fixed potential nil window access and corrected usage of strongSelf in blocks for grace timer and image tinting.

AdrianCurtin committed:
Improve scene and window handling for iOS/tvOS

Added tvOS 13.0 availability checks alongside iOS 13.0 for better platform compatibility. Improved logic to ensure mainWindow is only set if sceneDelegate.window exists, and made minor code style adjustments for consistency.

AdrianCurtin commented on SVProgressHUD/SVProgressHUD.m:
```diff
@@ -104,6 +134,9 @@ + (NSBundle *)imageBundle {
      NSBundle *bundle = [NSBundle bundleForClass:[SVProgressHUD class]];
 #endif
      NSURL *url = [bundle URLForResource:@"SVProgressHUD" withExtension:@"bundle"];
+     if (!url) {
+         return nil;
+     }
```
It also returned nil

AdrianCurtin commented:
@juliensaad 

This is a sort of bundled PR to close a bunch of open stability issues with the library, please review and comment as you see fit.

AdrianCurtin committed:
Update changelog for version 2.3.2

Added detailed entries for version 2.3.2, including fixes for iOS multi-scene window handling, CarPlay crashes, thread safety, and memory management improvements. Also noted tvOS 13.0 availability checks and clarified previous fixes.

PasqualePuzio commented:
Any news on this PR? I think it would be great if conflicts with commit ece669b were resolved so that the other fixes would be finally available

PasqualePuzio commented:
@AdrianCurtin could you please explain how issue #1153 was fixed by this pull request? What was the root cause? Thanks

AdrianCurtin committed:
Merge remote-tracking branch 'upstream/master'

AdrianCurtin commented:
"return windowScene.windows.firstObject;" was returning the wrong window (or nil) on iOS 26. 

The old code grabbed windowScene.windows.firstObject, which on iOS 26's stricter scene lifecycle is often not the key window It could be a system window or one that's hidden. So the HUD was being added to a non-visible window. Now it checks window.isKeyWindow && !window.isHidden, with multiple fallback layers (inactive scenes, delegate windows) if the first pass doesn't find one. Some other stuff isn't guaranteed to be called from the main thread so its essentilally a race condition.

AdrianCurtin committed:
Filter scenes by application role

Restrict scene handling to application-role UIWindowScenes on iOS/tvOS 13+. When searching for the keyboard window and the active foreground window, the code now checks scene.session.role == UIWindowSceneSessionRoleApplication before iterating windows. This avoids picking up non-application scenes (e.g. system, external, or keyboard-only scenes) and makes window selection more robust on multi-scene environments. No behavioral changes for earlier OS versions.

AdrianCurtin commented on SVProgressHUD/SVProgressHUD.m:
```diff
@@ -1308,10 +1362,32 @@ - (SVProgressHUDStyle) defaultStyleResolvingAutomatic {
-            keyboardWindow = testWindow;
-            break;
+ 
+    // First, try to find the keyboard window in all connected scenes
+    if (@available(iOS 13.0, tvOS 13.0, *)) {
+        for (UIScene *scene in UIApplication.sharedApplication.connectedScenes) {
+            // only handle UIWindowScene
+            if ([scene isKindOfClass:[UIWindowScene class]]) {
+                for (UIWindow *testWindow in ((UIWindowScene *)scene).windows) {
+                    if(![testWindow.class isEqual:UIWindow.class]) {
+                        keyboardWindow = testWindow;
+                        break;
+                    }
+                }
+            }
+            if (keyboardWindow != nil) {
+                break;
+            }
+        }
```
Addressed

AdrianCurtin commented on SVProgressHUD/SVProgressHUD.m:
```diff
@@ -1340,14 +1416,34 @@ - (CGFloat)visibleKeyboardHeight {
+    // For iOS 13 and later, we first find the active scene.
+    if (@available(iOS 13.0, tvOS 13.0, *)) {
+        for (UIScene *scene in UIApplication.sharedApplication.connectedScenes) {
+            if (scene.activationState == UISceneActivationStateForegroundActive) {
+                if ([scene isKindOfClass:[UIWindowScene class]]) {
+                    UIWindowScene *windowScene = (UIWindowScene *)scene;
+                    for (UIWindow *window in windowScene.windows) {
+                        // The isKeyWindow property is often a reliable way to find the main window,
+                        // but we also check other properties for robustness.
+                        if (window.isKeyWindow && window.alpha > 0) {
+                            return window;
+                        }
+                    }
+                }
+            }
+        }
```
Addressed
